components/mdns/mdns_browser.c

```diff
@@ -0,0 +1,630 @@
+        if (result->ttl == 0) {
+            queueDetach(mdns_result_t, browse->result, result);
+            // Just free current result
+            result->next = NULL;
+            mdns_query_results_free(result);
+        }
+        sync_result = sync_result->next;
+    }
+}
+
+/**
+ * @brief  Send PTR query packet to all available interfaces for browsing.
+ */
+static void _mdns_browse_send(mdns_browse_t *browse, mdns_if_t interface)
+{
+    // Using search once for sending the PTR query
+    mdns_search_once_t search = {0};
+
+    search.instance = NULL;
+    search.service = browse->service;
+    search.proto = browse->proto;
+    search.type = MDNS_TYPE_PTR;
+    search.unicast = false;
+    search.result = NULL;
+    search.next = NULL;
+
+    for (uint8_t protocol_idx = 0; protocol_idx < MDNS_IP_PROTOCOL_MAX; protocol_idx++) {
+        _mdns_search_send_pcb(&search, interface, (mdns_ip_protocol_t)protocol_idx);
+    }
+}
+
+void mdns_browse_send_all(mdns_if_t mdns_if)
+{
+    mdns_browse_t *browse = s_browse;
+    while (browse) {
+        _mdns_browse_send(browse, mdns_if);
+        browse = browse->next;
+    }
+}
+
+void mdns_browse_free(void)
+{
+    while (s_browse) {
+        mdns_browse_t *b = s_browse;
+        s_browse = s_browse->next;
+        _mdns_browse_item_free(b);
+    }
+}
+
+/**
+ * @brief  Mark browse as finished, remove and free it from browse chain
+ */
+static void _mdns_browse_finish(mdns_browse_t *browse)
+{
+    browse->state = BROWSE_OFF;
+    mdns_browse_t *b = s_browse;
+    mdns_browse_t *target_free = NULL;
+    while (b) {
+        if (strlen(b->service) == strlen(browse->service) && memcmp(b->service, browse->service, strlen(b->service)) == 0 &&
+                strlen(b->proto) == strlen(browse->proto) && memcmp(b->proto, browse->proto, strlen(b->proto)) == 0) {
+            target_free = b;
+            b = b->next;
+            queueDetach(mdns_browse_t, s_browse, target_free);
+            _mdns_browse_item_free(target_free);
+        } else {
+            b = b->next;
+        }
+    }
+    _mdns_browse_item_free(browse);
+}
+
+/**
+ * @brief  Allocate new browse structure
+ */
+static mdns_browse_t *_mdns_browse_init(const char *service, const char *proto, mdns_browse_notify_t notifier)
+{
+    mdns_browse_t *browse = (mdns_browse_t *)mdns_mem_malloc(sizeof(mdns_browse_t));
+
+    if (!browse) {
+        HOOK_MALLOC_FAILED;
+        return NULL;
+    }
+    memset(browse, 0, sizeof(mdns_browse_t));
+
+    browse->state = BROWSE_INIT;
+    if (!mdns_utils_str_null_or_empty(service)) {
+        browse->service = mdns_mem_strndup(service, MDNS_NAME_BUF_LEN - 1);
+        if (!browse->service) {
+            _mdns_browse_item_free(browse);
+            return NULL;
+        }
+    }
+
+    if (!mdns_utils_str_null_or_empty(proto)) {
+        browse->proto = mdns_mem_strndup(proto, MDNS_NAME_BUF_LEN - 1);
+        if (!browse->proto) {
+            _mdns_browse_item_free(browse);
```

Comment:
HOOK_MALLOC_FAILED.

I wonder that could we define some macro in the mdns_utils.h and using the same checking macro for the memory checking?

```
#define MDNS_MEM_CHECK(ptr, process) \
    do   {    \
        if (!ptr)    { \
            HOOK_MALLOC_FAILED;\
            process;   \
        }    \
     }
```

And use it like:
```
MDNS_MEM_CHECK(browse->proto, {_mdns_browse_item_free(browse);});
```
Just an idea, but `{_mdns_browse_item_free(browse);}` seems like odd...

Reply:
I'd like to stick with IDF style of error handling (basically esp_check.h), I'll handle this in phase III (error handling listed above, so I won't create any tracker)